### PR TITLE
Pluggable Storage

### LINKED ISSUE #492: Code Dev: Plugable Storage
We want to be able to have different storage for different environments. Because of that, we need a common input/output with each storage solution so Punchcard won't need to know which storage solution is being used.

``` gherkin
Feature: Plugable file storage
  As a core developer
  I want file storage options with common, asynchronous inputs and outputs
  So that I can have multiple storage options with the same interface.

  Scenario: Input
    Given that I will have a file or files which need to be stored
     When Punchcard sends the file or files to be stored
     Then it should include details on the file(s)
      And the details should include each file's name
      And the details should include each file's type
      And the details should include each file's location

  Scenario: Output with error
    Given that I gave the storage plugin a file which will cause errors
      And that those errors will happen when the storage plugin attempts to receive the file
     When the storage plugin receives an error when trying to store a file
     Then it should return an error object
      And the object should include a human-readable error string

  Scenario: Output with success
    Given that I gave the storage plugin a file which causes no errors
      And because it causes no errors, it is stored by the storage option that storage plugin called
     When storage plugin responds on successful storage of a file
     Then it should return a success object
      And that object should include a file path
      And that file path should be relative to the storage plugin's route.

  Scenario: Punchcard use when success
    Given that a user has submited a piece of content
      And that the content has no validation errors for any attributes
      And that said content included a file to be saved
      And that Punchcard has sent the file to the storage plugin
     When the storage plugin is successful in storing the file
     Then Punchcard receives a success response from the storage plugin
      And said response should include the relative file path for the file
      And Punchcard should save the piece of content to the database
      And said content should include the relative path to this file.

  Scenario: Punchcard use when failure
    Given that a user has submited a piece of content
      And that the content has no validation errors for any attributes
      And that said content included a file to be saved
      And that Punchcard has sent the file to the storage plugin
     When the storage plugin encounters an error when attempting to store the file
     Then Punchcard receives an error from the storage plugin
      And it should treat this error as a validation error
      And it should not save the data from the form
      And it should return the user to the data entry form
      And the data entry form should display the error from the plugin
      And it should display the error next to the file upload input
      And the error should look the same as other validation errors.
```

**storage plugin**
- [x] Input
- [ ] Error
- [x] Success
- [x] Punchcard good
- [ ] Punchcard bad

### PR DESCRIPTION
**Pull request by Snugug**
Add Pluggable Storage

---
Resolves #492
Replaces #528

`DCO 1.1 Signed-off-by: Sam Richard <sam@snug.ug>`

### CODE REVIEW

File: tests/fixtures/files/index.js [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+const uuid = require('uuid');
+const _ = require('lodash');
+const path = require('path');
+const fs = require('fs');
+
+const all = [
+  path.join(process.cwd(), '', 'storage.js'),
+  path.join(process.cwd(), '..', 'index.js'),
+  path.join(process.cwd(), '..', 'Gulpfile.js'),
+  path.join(process.cwd(), '../src/images', 'bee.svg'),
```
Comment (scottnath) [inline tests/fixtures/files/index.js]: Maybe not the `bee` image as a test file. Could def see that going away...

File: tests/storage.js [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    if (result === false) {
+      t.fail('No result for input item');
+    }
+
+    t.true(result.hasOwnProperty('name'), 'Result has a name');
+    t.true(result.hasOwnProperty('path'), 'Result has a path');
+    t.true(result.hasOwnProperty('type'), 'Result has a type');
+
+    t.is(result.name, expected.name, 'Result name is same as expected name');
+    t.is(result.type, expected.type, 'Result name is same as expected name');
```
Comment (scottnath) [inline tests/storage.js]: `Result type is same...`

File: lib/storage/put.js [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+      this._rel = pth;
+    }
+  }
+
+  static isCustomProp(name) {
+    return super.isCustomProp(name) && builtInProps.indexOf(name) === -1;
+  }
+}
+
+const mapFiles = file => {
```
Comment (scottnath) [inline lib/storage/put.js]: jsdoc

File: lib/storage/put.js [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+      f.type = file.type;
+      f.path = `/${f.uuid}${path.extname(file.originalFilename)}`;
+      f.rel = `${f.uuid}${path.extname(file.originalFilename)}`;
+
+      res(f);
+    });
+  });
+};
+
+const punchcardStream = items => {
```
Comment (scottnath) [inline lib/storage/put.js]: jsdoc

File: lib/storage/put.js [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+      .on('error', e => {
+        rej(e);
+      })
+      .on('end', () => {
+        res(files);
+      });
+  });
+};
+
+const put = files => {
```
Comment (scottnath) [inline lib/storage/put.js]: jsdoc

File: tests/fixtures/files/index.js [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+  path.join(process.cwd(), '../src/images', 'punchcard-image.svg'),
+  path.join(process.cwd(), '../src/images', 'punchcard-404.svg'),
+];
+
+const raw = () => {
```
Comment (scottnath) [inline tests/fixtures/files/index.js]: jsdoc

File: tests/fixtures/files/index.js [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+    });
+
+    files.splice(counter, 1);
+  } while (files.length > 0 && Math.random() > 0.5);
+
+
+  return items;
+};
+
+const saved = () => {
```
Comment (scottnath) [inline tests/fixtures/files/index.js]: jay es dock

File: config/default.js [hunk 8]
```diff
@@ -136,4 +136,10 @@ module.exports = {
       missing: 'Workflow \'%workflow\' for Content Type \'%type\' not found',
     },
   },
+  storage: {
+    type: 'fs',
```
Comment (Snugug) [inline config/default.js]: Easier said than done…
https://www.npmjs.com/search?q=vinyl

Comment (Snugug) [inline config/default.js]: This uses any [Vinyl Adapter](https://www.npmjs.com/package/vinyl#what-is-a-vinyl-adapter) that your heart can desire. There's no possible way we can document all the possibilities

Comment (scottnath) [inline config/default.js]: actually, you've started it just by having that link! Can link to that list, then explain a bit more about how to drop in one of them?

Comment (scottnath) [inline config/default.js]: Just promoting the idea of documenting this a little bit within this PR.

**Comment by Snugug** [comment]
@scottnath your docs, good sir

Harley cares greatly about good documentation

**Pull request merged**